### PR TITLE
fix(stepfunctions-tasks): incorrect S3 permissions for AthenaStartQueryExecution

### PR DESCRIPTION
**Pull request by Sumeet-Badyal**
The changes made by https://github.com/aws/aws-cdk/pull/11045 grant S3 scoped permissions if the optional parameter  output location is passed. The output location is not parsed correctly to be attached as a resource. When the output location is correctly parsed, state machines with a valid definition and a valid S3 bucket still fail due to an `Unable to verify/create output bucket` error. The exact same state machine and S3 bucket pass with wildcard permissions as such the resource for Start Query Execution must be changed to `*`.

BREAKING CHANGE: type of `outputLocation` in the experimental Athena `StartQueryExecution` has been changed to `s3.Location` from `string`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 1]
```diff
@@ -101,14 +101,14 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
           's3:ListBucketMultipartUploads',
           's3:ListMultipartUploadParts',
           's3:PutObject'],
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+        resources: ['*'], // State machines scoped to output location fail without * permissions  https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:104]: curious: is `*` needed on all of these APIs or does it apply to a subset of this list? if it's the latter, perhaps the policy statements should be broken up

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ describe('Start Query Execution', () => {
       },
       resultConfiguration: {
         encryptionConfiguration: { encryptionOption: EncryptionOption.S3_MANAGED },
-        outputLocation: 'https://s3.Region.amazonaws.com/bucket-name/key-name',
+        outputLocation: 's3://query-results-bucket/folder/',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:20]: is this change related to the changes being made in this PR? or is it just an unrelated improvement to update the input to something that is valid?

**Review by shivlaks — CHANGES_REQUESTED** [review]
@Sumeet-Badyal thanks for submitting the fix - I updated the commit message to indicate the error as per our `CONTRIBUTING` guide.

the rationale for that is because this will show up in the `CHANGELOG` under bug fixes and it reads better when it describes the problem that was fixed.

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts (hunk 2 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:20]: Update the input to something valid. The example in `lib/athena/start-query-execution.ts` is valid and should be used in the unit test.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 1 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts]: `s3:CreateBucket` was the only action that required wildcard permissions. This commit includes those changes 38cc8aa883a7b879fd3a7da708ffc21f1af5640c along with updates to the expected.json

Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts]: why not push a new policy statement with just createBucket scoped to `*`
i.e. - `policyStatements.push( new iam.PolicyStatement({...`

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts (hunk 2 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/athena/start-query-execution.test.ts:20]: got it, so it's not related to this change.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 1 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts]: This is out of date and the newest changes did that.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 3]
```diff
@@ -91,24 +91,30 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
         actions: ['s3:AbortMultipartUpload',
-          's3:CreateBucket',
           's3:GetBucketLocation',
           's3:GetObject',
           's3:ListBucket',
           's3:ListBucketMultipartUploads',
           's3:ListMultipartUploadParts',
           's3:PutObject'],
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+        resources: [this.props.resultConfiguration?.outputLocation ? 'arn:aws:s3:::' + this.props.resultConfiguration?.outputLocation.slice(5) + '*' : '*'], // Need S3 location where data is stored or Athena throws an Unable to verify/create output bucket https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:110]: should the type of s3 location be  `s3.Location` instead `https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.Location.html`- it would avoid having to do this `slice` and is more consistent with how s3 locations are used elsewhere within the CDK

Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:110]: Yes, will import and switch.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 4]
```diff
@@ -93,22 +94,28 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
     policyStatements.push(
       new iam.PolicyStatement({
-        actions: ['s3:AbortMultipartUpload',
-          's3:CreateBucket',
-          's3:GetBucketLocation',
-          's3:GetObject',
+        actions: ['s3:CreateBucket',
           's3:ListBucket',
+          's3:GetBucketLocation',
+          's3:GetObject'],
```
Comment (Sumeet-Badyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:100]: I had to add these actions for when one creates a S3 bucket on the console, then creates a state machine with the CDK with the output location set to this new S3 bucket, otherwise the state machine fails to get the object or bucket location.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 5]
```diff
@@ -93,22 +94,28 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+    );
+
+    policyStatements.push(
+      new iam.PolicyStatement({
+        actions: ['s3:AbortMultipartUpload',
           's3:ListBucketMultipartUploads',
           's3:ListMultipartUploadParts',
-          's3:PutObject'],
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+          's3:PutObject'], //s3://query-results-bucket/folder/
```
Comment (Sumeet-Badyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:110]: removing this comment in next commit

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 6]
```diff
@@ -93,22 +94,28 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+        resources: [this.props.resultConfiguration?.outputLocation?.bucketName ? 'arn:aws:s3:::' + this.props.resultConfiguration?.outputLocation?.bucketName + '/' + this.props.resultConfiguration?.outputLocation?.objectKey + '/*' : '*'], // Need S3 location where data is stored or Athena throws an Unable to verify/create output bucket https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
       }),
     );
 
     policyStatements.push(
       new iam.PolicyStatement({
         actions: ['lakeformation:GetDataAccess'],
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Workflow role permissions https://docs.aws.amazon.com/lake-formation/latest/dg/permissions-reference.html
+        resources: ['*'], // State machines scoped to output location fail and * permissions are required as per documentation https://docs.aws.amazon.com/lake-formation/latest/dg/permissions-reference.html
```
Comment (Sumeet-Badyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:118]: This resource does not get scoped to outputLocation. The only document I found that does not grant wildcard permissions is on page 78 of https://docs.aws.amazon.com/lake-formation/latest/dg/lake-formation-dg.pdf The default behaviour is to grant * access as found here: 
https://docs.aws.amazon.com/lake-formation/latest/dg/permissions-reference.html https://docs.aws.amazon.com/lake-formation/latest/dg/getting-started-setup.html
https://docs.aws.amazon.com/lake-formation/latest/dg/lake-formation-dg.pdf (page 21, 23, 27, 130, 150, 152, 153)

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 7]
```diff
@@ -167,25 +174,46 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
-        },
-        WorkGroup: this.props.workGroup,
-      }),
-    };
+          ResultConfiguration: {
+            EncryptionConfiguration: {
+              EncryptionOption: this.props.resultConfiguration?.encryptionConfiguration?.encryptionOption,
+              KmsKey: this.props.resultConfiguration?.encryptionConfiguration?.encryptionKey,
+            },
+            OutputLocation: 's3://' + this.props.resultConfiguration?.outputLocation?.bucketName + '/' + this.props.resultConfiguration?.outputLocation?.objectKey + '/',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:192]: prefer string interpolation to concatenation
```suggestion
            OutputLocation: `s3://${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/`,
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 4 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:100]: thanks for the heads up

**Review by shivlaks — CHANGES_REQUESTED** [review]
@Sumeet-Badyal I updated the commit description to include the breaking change entry so that it shows up in the changelog.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 8]
```diff
@@ -93,22 +94,28 @@ export class AthenaStartQueryExecution extends sfn.TaskStateBase {
...
+    );
+
+    policyStatements.push(
+      new iam.PolicyStatement({
+        actions: ['s3:AbortMultipartUpload',
           's3:ListBucketMultipartUploads',
           's3:ListMultipartUploadParts',
           's3:PutObject'],
-        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+        resources: [this.props.resultConfiguration?.outputLocation?.bucketName ? `arn:aws:s3:::${this.props.resultConfiguration?.outputLocation?.bucketName}/${this.props.resultConfiguration?.outputLocation?.objectKey}/*` : '*'], // Need S3 location where data is stored or Athena throws an Unable to verify/create output bucket https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
```
Comment (Sumeet-Badyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:111]: Changed this to String interpolation as well to follow the precedence at line 192

**Review by shivlaks — APPROVED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 8 above)
Comment (akiym) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:111]: @Sumeet-Badyal Thank you for your PR!
I've tried the changes in this PR, but since `objectKey` is required, it doesn't seem to work when specifying into directly under the bucket.
Another problem might be that when using [CREATE TABLE AS](https://docs.aws.amazon.com/athena/latest/ug/create-table-as.html), files are created under `tables/` directory of the bucket specified with `outputLocation`, so we have to specify `arn:aws:s3:::my-bucket/*` (not including any object keys except wildcard) instead of `arn:aws:s3:::my-bucket/foo/*`.
